### PR TITLE
Prepare everything to support profile as a new loadable mini-app

### PR DESCRIPTION
**Pull request by acelaya**
This PR has been split and is superseeded by:

* [User profile initial config](https://github.com/hypothesis/client/pull/5215) (*ready to review*)
  * └── [Add 'Your profile' item to UserMenu]() (*ready to review*)
    *  └── [User profile modal](https://github.com/hypothesis/client/pull/5217) (*draft*)

Separated reviews can be performed there.

---

This adds a new optional entry in the `UserMenu`, and prepares all the configurations and docs for the upcoming profile mini-app.

For now only the minimum logic required to open the new mini-app has been implemented. Once opened, it just loads an empty component.

image

### Test steps:

The new entry point is displayed in the menu only for LMS users that have the new `client_user_profile` flag enabled, so this needs to be tested from an LMS.

1. Checkout to this branch.
1. Checkout to the `profile-mini-app` branch in your local instance of `h`.
1. Login with an admin account (like `devdata_admin`), then go to http://localhost:5000/admin/features and enable `client_user_profile` (for "everyone" is ok).
1. Open an assignment on the LMS of your choice.
1. Open the sidebar and log-in if needed.
1. Click on the user icon. The menu should include the "Open profile" option.
1. Click on it to open the profile modal.
1. It should also be possible to close the profile modal and get back to the sidebar.
1. The menu entry should not be displayed for users without `client_user_profile` enabled, even in an LMS.
1. The menu entry should not be displayed outside of an LMS.

### Todo:

- [ ] Add basic tests for `profile.ts`, so that we can later potentially refactor it confidently.
- [ ] Add basic tests for `ProfileModal`, so that we can later refactor it confidently.

### Not covered

Right know the possible user flows are:

* Sidebar is open -> The user opens the notebook, which hides the sidebar.
* Notebook is open -> The user closes it, which hides the notebook and shows the sidebar.
* Sidebar is open -> The user opens the profile, which hides the sidebar.
* Profile is open -> The user closes it, which hides the profile and shows the sidebar.

I have not covered what would happen if the notebook is open and the user opens the profile, or the other way around, because those actions are not possible via the standard UI, and would introduce a lot of complexity.

### CODE REVIEW

File: docs/publishers/config.rst
Comment (lyzadanger) [inline docs/publishers/config.rst:481]: Well done to spot the need for this change!

**Review by lyzadanger — COMMENTED** [review]
Good start here. My thinking is that instead of having an _additional_ user-menu item when the associated feature flag is enabled, we should _replace_ the `Account settings` handler. The possible states are:

* First party user, flag not enabled: Behavior is unchanged (show account-settings option that links to `h` service)
* Third party user, flag not enabled: Behavior unchanged (no menu entry present for account settings)
* First party user, flag enabled: `Account settings` menu item should open profile view instead of linking to `h`
* Third party user, flag enabled: `Account settings` menu item should be present, and should open profile view

_Updated_: We can always change this logic later fairly easily.

File: src/sidebar/components/UserMenu.tsx
Comment (acelaya) [inline src/sidebar/components/UserMenu.tsx:49]: I was wondering if we want this item to depend on the fact that the user belongs to a third party.

Ideally I would not bind this option to that condition, and just make it depend on the feature flag, however, I don't know if we can make it show only to LMS users (for now) only based on that. If we can't, we might need to uncomment `/* isThirdParty || */` for now.

**Comment by acelaya** [comment]
> My thinking is that instead of having an additional user-menu item when the associated feature flag is enabled, we should replace the Account settings handler

That makes sense. In fact that was my initial idea, but I was trying to organize everything in my head and ended up "forgetting" that current settings need to be replaced. It does not make sense that a user could potentially end up seeing both.

@lyzadanger Considering it has a relatively high amount of conditions to display one thing or another/redirect to one place or another/behave on way or another, it might make sense to extract it to its own component (let's say `UserProfileMenuItem`). Would you agree with this?

**Comment by lyzadanger** [comment]
@acelaya As we discussed in a Slack conversation, I'm leaning toward agreeing with your first proposal that results in the following behavior:

* First-party users see the same `Account settings` MenuItem they've been seeing
* Third-party users _with the feature flag set_ see a new MenuItem that opens the profile mini-app

Whether or not the existing `Account settings` behavior and the new one are handled by the _same_ `MenuItem`: I'll leave that up to you! Perhaps you're right that there is more flexibility in making them separate `MenuItem`s...

**Comment by lyzadanger** [comment]
> The logic does not do anything just yet. I will prepare a follow-up PR which will provide the actual components for the profile app, and the communication between the host frame and the new profile one (or maybe I split that in two. Still to be defined).

Would you be amenable to chaining PRs a bit? While I do believe this PR is on the right track, it's hard to assess it without visible functionality. If you could get a chained PR (even in draft) of a basic `ProfileView` (or whatever) component that "opens" when you click the associated `MenuItem`, that would help us ensure this PR does what it's supposed to. Does that make sense?

In summary: This PR looks right, but I'd like to see the functionality that depends on it before we land it.

**Comment by acelaya** [comment]
> > The logic does not do anything just yet. I will prepare a follow-up PR which will provide the actual components for the profile app, and the communication between the host frame and the new profile one (or maybe I split that in two. Still to be defined).
> 
> Would you be amenable to chaining PRs a bit? While I do believe this PR is on the right track, it's hard to assess it without visible functionality. If you could get a chained PR (even in draft) of a basic `ProfileView` (or whatever) component that "opens" when you click the associated `MenuItem`, that would help us ensure this PR does what it's supposed to. Does that make sense?
> 
> In summary: This PR looks right, but I'd like to see the functionality that depends on it before we land it.

Absolutely agree. I was worried that the PR ended up being too big, and since I was still investigating what was needed for the next steps, I thought it was worth pushing the work already done to get some feedback and make sure I was moving in the right direction.

I'll push  acouple more changes so that there's actual new behavior, and then ping you for an extra review.

**Comment by lyzadanger** [comment]
> I'll push acouple more changes so that there's actual new behavior, and then ping you for an extra review.

Ah! To clarify, I was thinking that this PR is fine as it is, and perhaps you could open a _draft_ second PR that depends on this PR's branch. That way you don't have to worry about polishing the behavior (in that second PR), but we'll be able to evaluate that _this_ PR does what it should.

**Comment by acelaya** [comment]
Oooohh. Ok, I missunderstood you, hehe.

File: src/sidebar/components/ProfileView.tsx
Comment (acelaya) [inline src/sidebar/components/ProfileView.tsx:3]: Implementation to come on a follow-up PR

File: src/annotator/components/ProfileModal.tsx
Comment (acelaya) [inline src/annotator/components/ProfileModal.tsx:95]: This code is mostly duplicating what we have in `NotebookModal`.

On a follow-up PR I will try to unify the common stuff (almost everything) on a component that both ProfileModal and NotbookModal can wrap, and see if it does not make it more complex to understand.

File: src/annotator/config/index.js
Comment (acelaya) [inline src/annotator/config/index.js:67]: I still don't know what else will be needed for the profile. To be determined on the separated PR.

File: src/annotator/profile.tsx
Comment (acelaya) [inline src/annotator/profile.tsx:29]: This is mostly duplicating what's in `Notebook` class.

Maybe we could also unify it somehow, and then do something in the lines of:

```ts
const notebook = new WhateverDescriptiveAndClearName(...);
const profile = new WhateverDescriptiveAndClearName(...);
```

Instead of

```ts
const notebook = new Notebook(...);
const profile = new Profile(...);
```

But I have some doubts. It make make it less clear.

**Comment by acelaya** [comment]
> @acelaya As we discussed in a Slack conversation, I'm leaning toward agreeing with your first proposal that results in the following behavior:
> 
> * First-party users see the same `Account settings` MenuItem they've been seeing
> * Third-party users _with the feature flag set_ see a new MenuItem that opens the profile mini-app
> 
> Whether or not the existing `Account settings` behavior and the new one are handled by the _same_ `MenuItem`: I'll leave that up to you! Perhaps you're right that there is more flexibility in making them separate `MenuItem`s...

I have finally left two individual menu items, because it's easier to understand with less nested conditions. They are mutually exclussive anyway, as one requires the thirdParty condition to be true, and the other one requires it to be false.

**Comment by lyzadanger** [comment]
@acelaya NB: There is a missing testing step, which is to make sure you've enabled the associated feature flag locally at http://localhost:5000/admin/features

**Comment by acelaya** [comment]
> @acelaya NB: There is a missing testing step, which is to make sure you've enabled the associated feature flag locally at http://localhost:5000/admin/features

Good catch! I have updated the description.

**Review by lyzadanger — CHANGES_REQUESTED** [review]
Whew! There sure are a lot of places to touch, huh? We should keep an eye on that, especially if we ever find ourselves wanting to add yet another "mini-app." (We've already been talking about this).

The main bits and pieces of this PR are looking good. There are a few areas I spotted for a bit of further attention:

* I think there is an opportunity to remove some stuff we won't (immediately, at least) need from `ProfileModal`
* I'd encourage reviewing non-programmatic (comments, documentation) references to "profile" and qualify with "user" or otherwise make it more clear what "profile" means. It's a fairly abstract term on its own. I've left a few comments to this effect.
* I'm debating what the best initial label for the new `MenuItem` should be and welcome your thoughts. My gut feeling is that it should be just a noun, similar to how the current "Account settings" item is a noun (contrast with "Open notebook", which includes a verb; I recognize we have some inconsistency here). Ultimately, I think it will say "Account settings" but not initially. I don't think I have the right answer yet! Options I can think of include "Preferences", "Settings", "Your profile". Thoughts? I've also been ruminating about whether this menu would benefit from icons. Not to distract!
* I think there may be some opportunity to tighten up the factoring of some of the `UserMenu` tests for readability and brevity.

One important distinction I want to make is the difference between LMS and "Third-party". Your testing instructions indicate that the menu item would never be visible "outside of LMS" but that's not quite right: LMS is not the only third-party authority in `h`. So we need to be careful about that. There are more discussions to be had about which users should see this option and what should be in their respective user-profile views. More to think about!

What happens next I'll leave up to you! Options include:

* Break this PR into smaller pieces, if it is comfortable to do so. Some of the config-y bits should be easy to land, and then we can work through another pass or two of the component and app bits
* Move forward with this PR branch; convert out of draft. I anticipate another round or two of review before we'll be ready to land.

File: src/annotator/components/ProfileModal.tsx
Comment (acelaya) [reply src/annotator/components/ProfileModal.tsx:79]: Yep, that's the intention.

File: src/sidebar/components/UserMenu.tsx
Comment (acelaya) [reply src/sidebar/components/UserMenu.tsx:106]: I like the `Your profile` suggestion. It makes it more personal for the user to feel it as his/her own.

It seems to be a common practice.

image

File: src/annotator/profile.tsx
Comment (acelaya) [reply src/annotator/profile.tsx:29]: Agree

**Comment by acelaya** [comment]
> The main bits and pieces of this PR are looking good. There are a few areas I spotted for a bit of further attention:
> 
> * I think there is an opportunity to remove some stuff we won't (immediately, at least) need from `ProfileModal`

Done!

> * I'd encourage reviewing non-programmatic (comments, documentation) references to "profile" and qualify with "user" or otherwise make it more clear what "profile" means. It's a fairly abstract term on its own. I've left a few comments to this effect.

On it

> * I'm debating what the best initial label for the new `MenuItem` should be and welcome your thoughts. My gut feeling is that it should be just a noun, similar to how the current "Account settings" item is a noun (contrast with "Open notebook", which includes a verb; I recognize we have some inconsistency here). Ultimately, I think it will say "Account settings" but not initially. I don't think I have the right answer yet! Options I can think of include "Preferences", "Settings", "Your profile". Thoughts? I've also been ruminating about whether this menu would benefit from icons. Not to distract!

My "vote" goes for `Your profile`. I think it makes it pretty clear to the user what will be found there, and at the same time makes the user feel the apps is built to serve him/her.

> * I think there may be some opportunity to tighten up the factoring of some of the `UserMenu` tests for readability and brevity.

I'll give this another round of thought. However, do you mean specifically in the new tests I added, or in general?

> One important distinction I want to make is the difference between LMS and "Third-party". Your testing instructions indicate that the menu item would never be visible "outside of LMS" but that's not quite right: LMS is not the only third-party authority in `h`. So we need to be careful about that. There are more discussions to be had about which users should see this option and what should be in their respective user-profile views. More to think about!

Hmm, this is a good point.

Would you suggest we should do some change in code at this point, or just keep the `isThirdParty` condition until we have a better way to determine if first-party/LMS/other?

I will update the description anyway to better reflect this.

> What happens next I'll leave up to you! Options include:
> 
> * Break this PR into smaller pieces, if it is comfortable to do so. Some of the config-y bits should be easy to land, and then we can work through another pass or two of the component and app bits
> * Move forward with this PR branch; convert out of draft. I anticipate another round or two of review before we'll be ready to land.

I think it makes sense to split PRs, with a new set of better structured commits. I'll reference this PR from those.

**Comment by acelaya** [comment]
@lyzadanger I have split this pull request into 3 chained ones:

* [User profile initial config](https://github.com/hypothesis/client/pull/5215): It just adds the required configuration and documentation. There's no user-facing changis and should be safe to merge.
* [Add 'Your profile' item to UserMenu](https://github.com/hypothesis/client/pull/5214): Depends on the previous one. Only adds the entry point in the menu, and the logic to broadcast a message to the host frame.
* [User profile modal](https://github.com/hypothesis/client/pull/5217): Depends on the previous one. Adds the modal where the user profile mini-app will be displayed, and handles the message introduced in previous PR so that this modal can be displayed.

**Pull request closed without merging**
